nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/CopyAzureBlobStorage_v12.java

```diff
@@ -0,0 +1,330 @@
+            .name("copy-from-storage-credentials-service")
+            .displayName("Copy From Storage Credentials")
+            .description("Controller Service used to obtain Azure Blob Storage Credentials to read blob data ")
+            .identifiesControllerService(AzureStorageCredentialsService_v12.class)
+            .required(true)
+            .build();
+
+    public static final PropertyDescriptor COPY_FROM_CONTAINER = new PropertyDescriptor.Builder()
+            .name("copy-from-container-name")
+            .displayName("Copy From Container Name")
+            .description("If specified, copy from the specific container instead of using FlowFile content.")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .required(false)
+            .dependsOn(COPY_FROM_STORAGE_CREDENTIALS_SERVICE)
+            .build();
+
+    public static final PropertyDescriptor COPY_FROM_BLOB_NAME = new PropertyDescriptor.Builder()
+            .name("copy-from-blob-name")
+            .displayName("Copy From Blob Name")
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .required(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .description("If specified, copy from the specified blob name instead of using FlowFile content.")
+            .dependsOn(COPY_FROM_STORAGE_CREDENTIALS_SERVICE)
+            .build();
+
+    private static final List<PropertyDescriptor> PROPERTIES = Collections.unmodifiableList(Arrays.asList(
+            STORAGE_CREDENTIALS_SERVICE,
+            AzureStorageUtils.CONTAINER,
+            CREATE_CONTAINER,
+            CONFLICT_RESOLUTION,
+            BLOB_NAME,
+            COPY_FROM_STORAGE_CREDENTIALS_SERVICE,
+            COPY_FROM_CONTAINER,
+            COPY_FROM_BLOB_NAME,
+            AzureStorageUtils.PROXY_CONFIGURATION_SERVICE
+    ));
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    private BlobServiceClient fromStorageClient;
+
+    @OnScheduled
+    public void onScheduled(ProcessContext context) {
+        super.onScheduled(context);
+        final AzureStorageCredentialsService_v12 credentialsService = getCopyFromCredentialsService(context);
+        if (credentialsService != null) {
+            final HttpClient nettyClient = createHttpClient(context);
+            fromStorageClient = createFromStorageClient(nettyClient, credentialsService);
+        }
+    }
+
+    @OnStopped
+    public void onStopped() {
+        super.onStopped();
+        fromStorageClient = null;
+    }
+
+    private BlobServiceClient createFromStorageClient(final HttpClient nettyClient, final AzureStorageCredentialsService_v12 credentialsService) {
```

Comment:
I think this method is not needed and you can call `createStorageClient()` directly.